### PR TITLE
Clarify key discovery, remove OHTTP dependency

### LINKED ISSUE #25: Describe how origins specify which issuer to use
We need to describe how origins specify which issuer to use during an HTTP challenge

### LINKED ISSUE #33: Look at replacing OHTTP with a single encrypted header
Rather than using OHTTP, the client could send one header through the mediator to the issuer that is the origin ID, encrypted with the issuer's public key.

If the redeemer_name is empty in the token struct, then the origin ID header can be removed.

### LINKED ISSUE #36: Is the public key per issuer or per origin?

### PR DESCRIPTION
**Pull request by tfpauly**
Closes #25
Closes #33
Closes #36

- Use "origin" instead of "redeemer" for consistency
- Send the public keys in the challenge
- Validate the public keys on the mediator
- Remove OHTTP dependency, use a key for encrypting the ORIGIN_NAME instead

### CODE REVIEW

File: draft-private-access-tokens.md [hunk 1]
```diff
@@ -344,37 +369,26 @@ ORIGIN_NAME. An example request is shown below.
...
 ~~~
 
 Upon receipt, the Mediator computes ANON_CLIENT_ID, which is a fixed-length byte string
 for the given Client. See {{client-id}} for details of its computation. The Mediator
 also computes ANON_ORIGIN_ID_PRIME, a fixed-length byte string, for each ANON_ORIGIN_ID
 for a specific ANON_CLIENT_ID. See {{origin-id}} for details its computation.
 
-Before forwarding the Client's encapsulated request to the Issuer, the Mediator includes headers
+The Mediator also checks to validate that the key_id in the client's AccessTokenRequest
+matches a known public key for the Issuer. For example, the Mediator can fetch this
```
Comment (dvorak42) [inline draft-private-access-tokens.md:385]: I'm not sure if this is sufficient? Wouldn't the client be receiving keys immediately before the mediator checks the key against the issuer? (so the issuer can present key A to the client and for the next N minutes present key A to anyone who asks, before rotating to another key?)

**Review by chris-wood — APPROVED** [review]

File: draft-private-access-tokens.md [hunk 2]
```diff
@@ -208,24 +204,20 @@ issuance flow ({{issuance}}).
...
-Token redemption is an interactive protocol. Redeemers challenge Clients to
-present a unique, single-use token. Redeemers present this challenge to Clients
+Token redemption is an interactive protocol. Origins challenge Clients to
+present a unique, single-use token. Origins present this challenge to Clients
 with the following challenge:
 
 ~~~
 struct {
-   opaque redeemer_name<1..2^16-1>;
+   opaque origin_name<1..2^16-1>;
```
Comment (tfpauly) [inline draft-private-access-tokens.md:213]: Need to add a version field here

File: draft-private-access-tokens.md (hunk 1 above)
Comment (tfpauly) [reply draft-private-access-tokens.md:385]: Added some text at the end of the paragraph to address this

File: draft-private-access-tokens.md [hunk 3]
```diff
@@ -281,34 +290,34 @@ scheme defines one parameter, "token", which contains the base64url-encoded
...
+- Origin name (ORIGIN_NAME), a URI referring to the Origin {{!RFC6454}}. This is
+  the value of TokenChallenge.origin_name.
+- Issuer token public key (ISSUER_TOKEN_KEY), a blind signature public key
+  corresponding to the Issuer identified by TokenChallenge.issuer_name.
+- Issuer name public key (ISSUER_NAME_KEY), a public key used to encrypt requests
+  corresponding to the Issuer identified by TokenChallenge.issuer_name.
 
-Issuance also assumes that Issuers maintain local state for each distinct (redeemer, policy)
+Issuance also assumes that Issuers maintain local state for each distinct Origin
 tuple. In particular, for each tuple, Issuers maintain a stable mapping from ANON_CLIENT_ID
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-private-access-tokens.md]: Fixed this by adding more detail

File: draft-private-access-tokens.md [hunk 4]
```diff
@@ -235,12 +227,29 @@ redemption_nonce:
...
+- "token-key", which contains a base64url encoding of the SubjectPublicKeyInfo object
+for use with the RSA Blind Signature protocol (ISSUER_TOKEN_KEY).
+
+- "name-key", which contains a base64url encoding of the SubjectPublicKeyInfo object
+to use when encrypting the ORIGIN_NAME in issuance requests (ISSUER_NAME_KEY).
+
+- "max-age", an optional attribute that consists of the number of seconds for which
+the challenge will be accepted by the Origin.
+
+TODO: Define "name-key" more correctly.
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-private-access-tokens.md]: Removed =)

File: draft-private-access-tokens.md [hunk 5]
```diff
@@ -235,12 +227,29 @@ redemption_nonce:
...
+scheme uses the following attributes:
+
+- "challenge", which contains a base64url-encoded {{!RFC4648}} TokenChallenge
+value. This MUST be unique for every 401 HTTP response to prevent replay attacks.
+
+- "token-key", which contains a base64url encoding of the SubjectPublicKeyInfo object
+for use with the RSA Blind Signature protocol (ISSUER_TOKEN_KEY).
+
+- "name-key", which contains a base64url encoding of the SubjectPublicKeyInfo object
+to use when encrypting the ORIGIN_NAME in issuance requests (ISSUER_NAME_KEY).
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-private-access-tokens.md]: Applied

File: draft-private-access-tokens.md [hunk 6]
```diff
@@ -327,14 +336,30 @@ request using blinded_req:
...
+- "blinded_req" is the Nk-octet request defined above.
+
+The Client then generates an HTTP POST request to send through the Mediator to
+the Issuer, with the AccessTokenRequest as the body. The media type for this request
+is "message/access-token-request". The Client includes the "Token-Origin" header in
+this request, whose value is ENCRYPTED_ORIGIN_NAME; and the "Anonymous-Origin-ID" header,
+whose value is ANON_ORIGIN_ID. The Client sends this request to the
+Mediator's proxy URI. An example request is shown below, where Nk = 512.
+
+TODO: Describe how the origin name is encrypted
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-private-access-tokens.md]: Applied. Thank you so much!

File: draft-private-access-tokens.md (hunk 2 above)
Comment (tfpauly) [reply draft-private-access-tokens.md]: Added version.

**Comment by tfpauly** [comment]
@smhendrickson @dvorak42 please check, I'll plan on merging tomorrow

File: draft-private-access-tokens.md [hunk 7]
```diff
@@ -210,20 +210,22 @@ with the following challenge:
...
+    opaque issuer_name<1..2^16-1>;
+    opaque redemption_nonce[32];
 } TokenChallenge;
 ~~~
 
-origin_name:
-: Name of the Origin (ORIGIN_NAME).
+The structure fields are defined as follows:
+
+- "version" is a 1-octet integer. This document defines version 1.
```
Comment (dvorak42) [inline draft-private-access-tokens.md:222]: Maybe somewhere we should toss some wording that consumers should not attempt to parse unknown versions, to try avoiding people just ignoring the version, trying to parse everything else and exploding (*cough* TLS *cough*).

File: draft-private-access-tokens.md [hunk 8]
```diff
@@ -312,8 +314,8 @@ Issuance assumes the Client has the following information, derived from a given
 - Issuer name public key (ISSUER_NAME_KEY), a public key used to encrypt requests
   corresponding to the Issuer identified by TokenChallenge.issuer_name.
 
-Issuance also assumes that Issuers maintain local state for each distinct Origin
-tuple. In particular, for each tuple, Issuers maintain a stable mapping from ANON_CLIENT_ID
+Issuance also assumes that Issuers maintain local state for each distinct pair of Client
+and Origin. In particular, for pair, Issuers maintain a stable mapping from ANON_CLIENT_ID
```
Comment (dvorak42) [inline draft-private-access-tokens.md:318]: nit: for each pair

File: draft-private-access-tokens.md [hunk 9]
```diff
@@ -476,6 +479,43 @@ sig = rsabssa_finalize(ISSUER_TOKEN_KEY, nonce, blind_sig, blind_inv)
...
+1. Encrypt (seal) request with aad as associated data using context, yielding ciphertext ct.
+1. Concatenate the values of aad, enc, and ct, yielding an Encapsulated Request enc_request.
+
+Note that enc is of fixed-length, so there is no ambiguity in parsing this structure.
+
+In pseudocode, this procedure is as follows:
+
+~~~
+enc, context = SetupBaseS(pkR, "OriginTokenRequest")
+aad = concat(encode(1, keyID),
```
Comment (dvorak42) [inline draft-private-access-tokens.md:507]: Does this need a prefix?

Comment (dvorak42) [reply draft-private-access-tokens.md:507]: nevermind, I guess the SetupBaseS info is sufficient

**Review by dvorak42 — APPROVED** [review]

File: draft-private-access-tokens.md [hunk 10]
```diff
@@ -385,65 +402,120 @@ listing both the ANON_CLIENT_ID as "Anonymous-Client-ID", and the ANON_ORIGIN_ID
...
+- The decrypted ORIGIN_NAME matches an Origin the Issuer serves
+
+If any of these conditions is not met, the Issuer MUST return an HTTP 400 error to the Mediator,
+which will forward the error to the client.
+
+If the conditions are met, the Issuer then applies its policy for the Client request, by checking
+for state associated with the ANON_CLIENT_ID. If there is previous state associated with the
+ANON_CLIENT_ID, the Issuer first validates two conditions:
+
+- The decrypted ORIGIN_NAME for this request matches the ORIGIN_NAME stored from previous requests
```
Comment (smhendrickson) [inline draft-private-access-tokens.md:433]: Should this instead just check for the unique tuple of {ANON_CLIENT_ID, ORIGIN_NAME}? A collision is highly unlikely, but I think this makes the case less confusing.

File: draft-private-access-tokens.md (hunk 8 above)
Comment (tfpauly) [reply draft-private-access-tokens.md]: Fixed!

File: draft-private-access-tokens.md (hunk 7 above)
Comment (tfpauly) [reply draft-private-access-tokens.md:222]: Good call. I've added text in all of the places that use version to explain how it is verified and must gate parsing/usage.

**Review by janaiyengar — COMMENTED** [review]
Thanks, @tfpauly! Apologies for the late review, but LGTM overall, with one minor nit. I will do a full read through again of the mechanism to ensure that it sits well.

File: draft-private-access-tokens.md [hunk 11]
```diff
@@ -344,37 +373,32 @@ ORIGIN_NAME. An example request is shown below.
...
-Before forwarding the Client's encapsulated request to the Issuer, the Mediator includes headers
+The Mediator also checks to validate that the key_id in the client's AccessTokenRequest
+matches a known public key for the Issuer. For example, the Mediator can fetch this
+key using the API defined in {{setup}}. This check is done to help ensure that the Client
+has not been given a unique key that could allow the Issuer to fingerprint or target
+the Client. If the key does not match, the Mediator rejects the request with an HTTP
+400 error. Note that Mediators need to be careful in cases of key rotation; particularly,
+an Origin could collude with an Issuer to try to rotate the key for each new Client in
+order to link the client activity. The policy Mediators apply for key validation ought
+to take such attacks into consideration.
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-private-access-tokens.md:396]: Agreed! Filed #44.

**Pull request merged**